File: 2023/01/python/main.py
#! /usr/bin/env python3

""" Day one of Advent of code."""

from argparse import ArgumentParser, FileType

DIGITS_STR = ['one', 'two', 'three', 'four', 'five', 'six', 'seven', 'eight',
              'nine']


def get_digits(line, part_two):
    """Return the digits encountered in a given line."""
    ret = []
    for i, x in enumerate(line):
        if '0' <= x <= '9':
            ret.append(ord(x) - ord('0'))
        elif part_two:
            for j, y in enumerate(DIGITS_STR):
                if line[i:].startswith(y):
                    ret.append(j + 1)

    return ret


def compute_line(line, part_two):
    """Process a line extracting first and last digit."""
    digits = get_digits(line, part_two)
    return digits[0] * 10 + digits[-1]


def __main__():
    parser = ArgumentParser(description=__doc__)

    parser.add_argument(dest='input', type=FileType(encoding='utf-8'))
    parser.add_argument('--part-two', action='store_true')

    args = parser.parse_args()

    print(sum(compute_line(line, args.part_two) for line in args.input))


if __name__ == '__main__':
    __main__()

File: 2023/01/rust/src/main.rs
use std::fs::File;
use std::io::{self, BufRead};
use std::path::Path;


// https://doc.rust-lang.org/rust-by-example/std_misc/file/read_lines.html
fn read_lines<P>(filename: P) -> io::Result<io::Lines<io::BufReader<File>>>
where P: AsRef<Path>, {
    let file = File::open(filename)?;
    Ok(io::BufReader::new(file).lines())
}

static DIGITS: [&str ; 9] = [
    "one", "two", "three", "four", "five", "six", "seven", "eight", "nine"
];

fn is_str_digit(s: &str) -> Option<u32> {
    for (i, x) in DIGITS.iter().enumerate() {
        if s.starts_with(x) {
            return Some (i as u32 + 1)
        }
    }
    None
}

fn is_ascii_digit(c: char) -> Option<u32> {
    c.to_digit(10)
}

fn get_digits(line: &str) -> Vec<u32> {
    let mut ret:Vec<u32> = Vec::new();

    for (i, c) in line.chars().enumerate() {
        if let Some(b) = is_ascii_digit(c).or(is_str_digit(&line[i..])) {
            ret.push(b);
        }
    }

    ret
}

fn main() -> io::Result<()> {
    let input_path = std::env::args().nth(1).expect("No input file given.");

    let lines = read_lines(input_path).unwrap();

    let mut ret: Vec<u32> = Vec::new();

    for line in lines {
        let line = line?;
        let digits = get_digits(&line);

        ret.push(digits[0] * 10 + digits.last().unwrap());
    }

    let total: u32 = ret.iter().sum();
    println!("{}", total);

    Ok(())
}
